src/rpc.rs

```diff
@@ -1518,6 +1518,75 @@ impl <VatId> ConnectionState<VatId> {
     }
 }
 
+enum DisconnectorState
+{
+    Connected,
+    Disconnecting,
+    Disconnected
+}
+
+pub struct Disconnector<VatId> where VatId: 'static {
+    connection_state: Rc<RefCell<Option<Rc<ConnectionState<VatId>>>>>,
+    state:  DisconnectorState,
+}
+
+impl <VatId> Disconnector<VatId> {
+    pub fn new(connection_state: Rc<RefCell<Option<Rc<ConnectionState<VatId>>>>>) -> Disconnector<VatId> {
+        let state = match *(connection_state.borrow()) {
+            Some(_) => {
+                DisconnectorState::Connected
+            },
+            None => {
+                DisconnectorState::Disconnected
+            }
+        };
+        Disconnector {
+            connection_state: connection_state,
+            state: state,
+        }
+    }
+    fn disconnect(&self) {
```

Comment:
Was this perhaps intended to be public? I'm having trouble figuring out the correct way to shut down the RPC system in response to an error on the transport layer.

Reply:
Hm... yeah, it does look to me like this was intended to be public.

> the correct way to shut down the RPC system in response to an error on the transport layer

Note that I/O errors on the underlying stream automatically trigger a disconnect and shut down the RPC system. If the I/O error is [something other than a `capnp::ErrorKind::Disconnected`](https://github.com/capnproto/capnproto-rust/blob/4e6859c74251897a44c222e9679eea74145f2e33/capnp-rpc/src/lib.rs#L196), then that error gets propagated to the driving `RpcSystem` future.

Do you have a situation where that's not true, or where that's not sufficient?

(The `RpcSystem` setup and teardown logic is kinda janky right now, so I wouldn't be surprised if it has some problems.)

Reply:
I'm not sure my current setup is something that's a priority to support:-)I have a TcpStream that talks to the network, and created two closures to implement a simple message format (16 bit length + payload) on top that, then encrypt/decrypt that using [Snow](https://github.com/mcginty/snow).

To give capnproto an AsyncRead/AsyncWrite, I create a UnixStream::pair with one end being given to capnproto and the other split to the encrypt/decrypt closures. I'd like to migrate this to implementing my own AsycnRead/AsyncWrite to avoid unnecessary trips to kernel space and platform-specific APIs, but that proved more complicated than I'd expected so I've been putting it off, and this mostly works except shutdown is challenging. The concept is similar to Tokio's frame codecs (which I also struggled with on a previous project some time ago).

I've used this trick a fair bit in Rust and other languages, and to get a clean shutdown, you call shutdown on all sockets (both ends of the pair and the network-facing socket) whenever there's an error detected on any socket. The problem I'm running into with Rust async_std is that it is less tolerant of my unsafely cloning the raw socket and then shutting it down from outside the executor in another thread, and panics when it can't remove the I/O watch which is honestly pretty understandable.

Anyway the right answer is for me to stop abusing Unix-specific syscalls to get around Rust ownership and async framework design, and implement AsyncRead/AsyncWrite or try tokio codecs again.

Reply:
I didn't make it public because it was intended to run as a Future after you were done with the connection.  see https://github.com/ian-p-cooke/capntls/blob/9b5eb98607184a05ef309203dc2ab98d352ac124/src/client.rs#L64 for my example.  That doesn't mean it -has- to be private I just didn't need it to be public for what I imagined.

Reply:
Thanks -- running it as a future makes sense, and should solve my problem.